avm/src/daemon/ctrl.rs

```diff
@@ -558,6 +567,413 @@ fn handle_keys() -> Result<Response<Body>, Infallible> {
+        res_hm.push_register(0, 1)?;
+      }
+    }
+    None => {
+      res_hm.push_fixed(0, 0);
+      res_hm.push_fractal(
+        0,
+        HandlerMemory::str_to_fractal("namespace-key pair not found"),
+      )?;
+    }
+  }
+  Ok(res_hm)
+}
+
+async fn handle_dsmrun(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+  match dsmrun_inner(req).await {
+    Ok(hand_mem) => {
+      let mut out = vec![];
+      hand_mem.to_pb().write_to_vec(&mut out).unwrap();
+      Ok(Response::builder().status(200).body(out.into()).unwrap())
+    }
+    Err(err) => {
+      // TODO: What error message here? Also should this also be a valid HM out of here?
+      eprintln!("{:?}", err);
+      Ok(Response::builder().status(500).body("fail".into()).unwrap())
+    }
+  }
+}
+
+async fn dsmrun_inner(req: Request<Body>) -> DaemonResult<Arc<HandlerMemory>> {
+  let headers = req.headers();
+  let nskey = headers
+    .get("nskey")
+    .map_or("N/A", |v| v.to_str().unwrap())
+    .to_string();
+  let maybe_hm = DS.get(&nskey);
+  let subhandler_id = headers
+    .get("subhandler_id")
+    .map_or(0, |v| v.to_str().unwrap().parse().unwrap());
+  let subhandler = HandlerFragment::new(subhandler_id, 0);
+  let bytes = body::to_bytes(req.into_body()).await?;
+  let pb = protos::HandlerMemory::HandlerMemory::parse_from_bytes(&bytes)?;
+  let mut hm = HandlerMemory::from_pb(&pb)?;
+  let mut res_hm = HandlerMemory::new(None, 1)?;
+  res_hm.init_fractal(0)?;
+  match maybe_hm {
+    Some(ds) => {
+      HandlerMemory::transfer(&ds, 0, &mut hm, CLOSURE_ARG_MEM_START + 1)?;
+      let hm = subhandler.run(hm).await?;
+      res_hm.push_fixed(0, 1);
+      if hm.addr_to_idxs_opt(CLOSURE_ARG_MEM_START).is_some() {
+        // Guard against void functions
+        HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START, &mut res_hm, 1);
+        res_hm.push_register(0, 1)?;
+      }
+      // Also grab the mutation to the datastore value and re-insert it
+      let mut newds = HandlerMemory::new(None, 1)?;
+      HandlerMemory::transfer(&hm, CLOSURE_ARG_MEM_START + 1, &mut newds, 0)?;
+      drop(ds);
+      DS.insert(nskey, newds);
+    }
+    None => {
+      res_hm.push_fixed(0, 0);
+      res_hm.push_fractal(
+        0,
+        HandlerMemory::str_to_fractal("namespace-key pair not found"),
+      )?;
+    }
+  }
+  Ok(res_hm)
+}
+
+async fn handle_dsrwith(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+  match dsrwith_inner(req).await {
+    Ok(hand_mem) => {
+      let mut out = vec![];
+      hand_mem.to_pb().write_to_vec(&mut out).unwrap();
+      Ok(Response::builder().status(200).body(out.into()).unwrap())
+    }
+    Err(err) => {
+      // TODO: What error message here? Also should this also be a valid HM out of here?
+      eprintln!("{:?}", err);
+      Ok(Response::builder().status(500).body("fail".into()).unwrap())
+    }
+  }
+}
+
+async fn dsrwith_inner(req: Request<Body>) -> DaemonResult<Arc<HandlerMemory>> {
```

Comment:
nit: `dsrwith_inner` is the exact same as `dsrrun_inner` - is there a reason they're not collapsed into a single function? from what i can tell, all of the `_inner` functions can be collapsed into 1 function with an enum, which can be passed into a `handle_distributed` which collapses all of the `handle_` functions

Reply:
i see the side-effect-only functions below - those can be handled with a `handle_distributed_side_effect` which would be nearly the same as `handle_distributed` except it would `.map(|_| ())` the `Result` from the collapsed `_inner` function